### ISSUE TITLE
[api-extractor] Different ordering in api properties with node 10 and node 12

### ISSUE THREAD

**Issue by KarishmaGhiya** [microsoft/rushstack#1552]
## [api-extractor] Different ordering in api properties with node 10 and node 12

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**
Different ordering of the api properties with node 10 and node 12.

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**
To reproduce this behaviour, checkout master at https://github.com/Azure/azure-sdk-for-js

Run these steps -
Install node 10
rush install
rush build --to @azure/event-hubs
Save the generated api output

Install node 12
rush install
rush build --to @azure/event-hubs
Save the generated api output

**What is the expected behavior?**
The two api outputs have different re-ordering of the properties. One suggested solution is to reorder the properties alphabetically so that reordering is consistent across different node versions.

Cc: @mikeharder

**Comment by octogonz** [comment]
The API members are sorted in [ApiItemContainer.members](https://github.com/microsoft/rushstack/blob/47e9194e61e6d065d69394d0e334c18b976de180/apps/api-extractor-model/src/mixins/ApiItemContainerMixin.ts#L146) based on [ApiItem.getSortKey()](https://github.com/microsoft/rushstack/blob/47e9194e61e6d065d69394d0e334c18b976de180/apps/api-extractor-model/src/items/ApiItem.ts#L259).  Not sure why the results would be different in node 10 vs 12.  

We'll need to step through it in a debugger to find out.

**Comment by octogonz** [comment]
@KarishmaGhiya Which output file were you referring to?  In your repro, the .d.ts rollups look the same, and the .api.json file is not enabled.  Thus the only difference I see is in the API report file, for this one constructor overload:

```ts
// @public
export class EventHubClient {
    constructor(connectionString: string, options?: EventHubClientOptions);
    constructor(connectionString: string, eventHubName: string, options?: EventHubClientOptions);
    constructor(host: string, eventHubName: string, credential: TokenCredential, options?: EventHubClientOptions);  <====
    close(): Promise<void>;
```
versus
```ts
// @public
export class EventHubClient {
    constructor(host: string, eventHubName: string, credential: TokenCredential, options?: EventHubClientOptions);  <====
    constructor(connectionString: string, options?: EventHubClientOptions);
    constructor(connectionString: string, eventHubName: string, options?: EventHubClientOptions);
    close(): Promise<void>;
```

**Comment by mikeharder** [comment]
@octogonz: I believe the different ordering of the constructor overloads is the exact issue we are reporting.  This ordering is different across Node 10 and Node 12.

**Comment by octogonz** [comment]
The API report uses [Collector.getSortKeyIgnoringUnderscore()](https://github.com/microsoft/rushstack/blob/18b34857f4b77ddc58c769c7b2aeffab61efadf0/apps/api-extractor/src/generators/ApiReportGenerator.ts#L290).

All three constructors get the same sort key value of `"_constructor*_constructor*_"`, which means they should preserve their original ordering from the source file, since `Array.prototype.sort()` is supposed to be a stable sort.

It turns out that NodeJS has a bug where `Array.prototype.sort()` is not stable if the array has more than 10 elements.  https://github.com/nodejs/node/issues/29446  

They fixed it in Node 12, but chose not to backport this fix to Node 10 or 8.

We should improve the sort key to include the `overloadIndex`.

**Comment by octogonz** [comment]
@KarishmaGhiya are you able to create a PR?

**Comment by KarishmaGhiya** [comment]
@octogonz I haven't got a chance to follow up on this yet. I will let you know after I create a PR. Thanks.

**Comment by octogonz** [comment]
BTW if you need some pointers, I'm reachable on [Gitter](https://gitter.im/octogonz).

**Comment by scsewall** [comment]
We are running into this same issue of different ordering in the api.md file between Node 10 and Node 12. In addition to overloaded constructors, we are also hitting it when methods or properties have the same name in a static and a non-static form. Probably, the same issue, but I will test when the fix is available.

**Comment by scsewall** [comment]
@KarishmaGhiya, @octogonz Any update on this issue? This is now preventing us from adding official Node 12 support (want to continue to support Node 10) as the unstable ordering in the api.md file would cause problems for our CI builds.

**Comment by octogonz** [comment]
If no one else is able to work on this, I can probably do it over the weekend. It's a pretty easy fix I suspect

**Comment by octogonz** [comment]
Here's a PR: https://github.com/microsoft/rushstack/pull/1613

I decided to do more fundamental fix, and provide a way to make all sorts stable.

**Comment by octogonz** [comment]
@mikeharder If you guys could test this PR branch against your repro, that might save me some time.

**Comment by mikeharder** [comment]
@octogonz: We will test this today.

**Comment by mikeharder** [comment]
@octogonz: In our repo, I have verified this change makes the Node 10 ordering match the existing Node 12 ordering.

**Comment by octogonz** [comment]
Published as API Extractor 7.5.4

**Comment by calebmshafer** [comment]
@scsewall is away at a conference this week, but I just verified with our repository and everything is working with v7.5.4.  Thanks!
